File: src/lib/riskEngine.ts
import type {
  RiskAnalysis,
  RiskLevel,
  RecommendedAction,
  RiskSignal,
  ThreatType,
  TimelineEvent,
  BehavioralFingerprint,
} from '@/types';

export interface RiskEngineInput {
  amount: number;
  currency?: string;
  account_age_days: number;
  is_new_device: boolean;
  is_new_location: boolean;
  failed_attempts: number;
  transaction_velocity: number;
  historical_average: number;
  transaction_id?: string;
  timestamp?: string;
  location?: string;
  device_info?: string;
}

export interface RiskEngineResult {
  risk_score: number;
  risk_level: RiskLevel;
  threat_type: ThreatType;
  recommended_action: RecommendedAction;
  risk_factors: RiskSignal[];
  attack_story: string;
  timeline: TimelineEvent[];
  behavioral_fingerprint: BehavioralFingerprint[];
}

const SIGNAL_WEIGHTS = {
  transaction_amount_anomaly: 22,
  transaction_velocity: 18,
  new_device: 14,
  new_location: 12,
  account_age: 10,
  failed_attempts: 14,
  historical_behavior_deviation: 10,
} as const;

const MAX_SCORE = Object.values(SIGNAL_WEIGHTS).reduce((a, b) => a + b, 0);

export function classifyRiskLevel(score: number): RiskLevel {
  if (score <= 30) return 'LOW';
  if (score <= 60) return 'MEDIUM';
  if (score <= 80) return 'HIGH';
  return 'CRITICAL';
}

export function recommendAction(level: RiskLevel): RecommendedAction {
  switch (level) {
    case 'LOW':
      return 'APPROVE';
    case 'MEDIUM':
      return 'REVIEW';
    case 'HIGH':
    case 'CRITICAL':
      return 'HOLD';
  }
}

function clamp(n: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, n));
}

function roundTo(n: number, decimals = 0): number {
  const f = 10 ** decimals;
  return Math.round(n * f) / f;
}

export function analyzeRisk(input: RiskEngineInput): RiskEngineResult {
  const signals: RiskSignal[] = [];

  // 1. Transaction amount anomaly
  const avg = input.historical_average || 1000;
  const amountRatio = input.amount / avg;
  const amountAnomalyScore = amountRatio > 1 ? clamp((amountRatio - 1) * 30, 0, 100) : 0;
  const amountTriggered = amountRatio > 2.5;
  signals.push({
    key: 'transaction_amount_anomaly',
    label: 'Transaction Amount Anomaly',
    weight: SIGNAL_WEIGHTS.transaction_amount_anomaly,
    triggered: amountTriggered,
    value: `${input.currency || '₹'}${input.amount.toLocaleString()}`,
    normal_value: `${input.currency || '₹'}${avg.toLocaleString()}`,
    deviation: roundTo(amountAnomalyScore, 1),
  });

  // 2. Transaction velocity
  const velocityScore = clamp(input.transaction_velocity * 12, 0, 100);
  const velocityTriggered = input.transaction_velocity >= 5;
  signals.push({
    key: 'transaction_velocity',
    label: 'Transaction Velocity',
    weight: SIGNAL_WEIGHTS.transaction_velocity,
    triggered: velocityTriggered,
    value: `${input.transaction_velocity} txns / 10 min`,
    normal_value: '0–2 txns / 10 min',
    deviation: roundTo(velocityScore, 1),
  });

  // 3. New device
  signals.push({
    key: 'new_device',
    label: 'New Device',
    weight: SIGNAL_WEIGHTS.new_device,
    triggered: input.is_new_device,
    value: input.is_new_device ? 'New device detected' : 'Known device',
    normal_value: 'Known device',
    deviation: input.is_new_device ? 100 : 0,
  });

  // 4. New location
  signals.push({
    key: 'new_location',
    label: 'New Location',
    weight: SIGNAL_WEIGHTS.new_location,
    triggered: input.is_new_location,
    value: input.is_new_location ? `${input.location || 'New location'}` : 'Known location',
    normal_value: 'Known location',
    deviation: input.is_new_location ? 100 : 0,
  });

  // 5. Account age
  const ageScore = clamp((30 - input.account_age_days) * 3.3, 0, 100);
  const ageTriggered = input.account_age_days < 7;
  signals.push({
    key: 'account_age',
    label: 'Account Age',
    weight: SIGNAL_WEIGHTS.account_age,
    triggered: ageTriggered,
    value: `${input.account_age_days} days`,
    normal_value: '> 90 days',
    deviation: roundTo(ageScore, 1),
  });

  // 6. Failed attempts
  const failedScore = clamp(input.failed_attempts * 20, 0, 100);
  const failedTriggered = input.failed_attempts >= 3;
  signals.push({
    key: 'failed_attempts',
    label: 'Failed Authentication Attempts',
    weight: SIGNAL_WEIGHTS.failed_attempts,
    triggered: failedTriggered,
    value: `${input.failed_attempts} attempts`,
    normal_value: '0 attempts',
    deviation: roundTo(failedScore, 1),
  });

  // 7. Historical behavior deviation (composite)
  const compositeDeviation =
    (amountAnomalyScore * 0.3 +
      velocityScore * 0.25 +
      (input.is_new_device ? 100 : 0) * 0.15 +
      (input.is_new_location ? 100 : 0) * 0.15 +
      ageScore * 0.05 +
      failedScore * 0.1) /
    1.0;
  const histTriggered = compositeDeviation > 50;
  signals.push({
    key: 'historical_behavior_deviation',
    label: 'Historical Behavior Deviation',
    weight: SIGNAL_WEIGHTS.historical_behavior_deviation,
    triggered: histTriggered,
    value: `${roundTo(compositeDeviation, 1)}% deviation`,
    normal_value: '< 20% deviation',
    deviation: roundTo(compositeDeviation, 1),
  });

  // Weighted score
  let rawScore = 0;
  for (const s of signals) {
    rawScore += (s.deviation / 100) * s.weight;
  }
  const risk_score = clamp(roundTo((rawScore / MAX_SCORE) * 100, 0), 0, 100);
  const risk_level = classifyRiskLevel(risk_score);
  const recommended_action = recommendAction(risk_level);

  // Threat classification
  const threat_type = classifyThreat(input, signals);

  // Attack story
  const attack_story = generateAttackStory(input, signals, threat_type);

  // Timeline
  const timeline = generateTimeline(input, risk_score);

  // Behavioral fingerprint
  const behavioral_fingerprint = generateFingerprint(input, signals);

  return {
    risk_score,
    risk_level,
    threat_type,
    recommended_action,
    risk_factors: signals,
    attack_story,
    timeline,
    behavioral_fingerprint,
  };
}

function classifyThreat(
  input: RiskEngineInput,
  signals: RiskSignal[]
): ThreatType {
  const newDevice = input.is_new_device;
  const newLocation = input.is_new_location;
  const failed = input.failed_attempts >= 3;
  const highVelocity = input.transaction_velocity >= 5;
  const youngAccount = input.account_age_days < 7;
  const highAmount = input.amount / (input.historical_average || 1000) > 3;

  const triggered = signals.filter((s) => s.triggered).map((s) => s.key);

  if (triggered.length === 0 || (triggered.length === 1 && triggered[0] === 'account_age' && !youngAccount)) {
    return 'Normal Transaction';
  }

  if (newDevice && newLocation && failed) return 'Possible Account Takeover';
  if (highVelocity && highAmount) return 'Velocity Attack';
  if (newDevice && !newLocation && !failed) return 'Suspicious Device';
  if (highAmount && !newDevice && !newLocation) return 'Unusual Amount';
  if (failed && !newDevice && !newLocation) return 'Multiple Failed Attempts';
  if (youngAccount && (newDevice || newLocation)) return 'Potential Synthetic Identity';

  if (newDevice && newLocation) return 'Possible Account Takeover';
  if (highVelocity) return 'Velocity Attack';
  if (newDevice) return 'Suspicious Device';

  return 'Unusual Amount';
}

function generateAttackStory(
  input: RiskEngineInput,
  signals: RiskSignal[],
  threat: ThreatType
): string {
  if (threat === 'Normal Transaction') {
    return 'This transaction appears consistent with the customer\'s historical behavior. No significant anomalies were detected across device, location, amount, or velocity signals. The risk engine recommends approval with standard monitoring.';
  }

  const parts: string[] = [];

  if (input.is_new_device) {
    parts.push('A previously unseen device was used to initiate this transaction');
  }
  if (input.is_new_location) {
    parts.push(`the transaction originated from a new geographic location (${input.location || 'unrecognized region'})`);
  }
  if (input.failed_attempts >= 3) {
    parts.push(`${input.failed_attempts} failed authentication attempts preceded the successful transaction`);
  }
  if (input.transaction_velocity >= 5) {
    parts.push(`${input.transaction_velocity} transactions were attempted within a short 10-minute window, significantly above the expected baseline`);
  }
  const ratio = input.amount / (input.historical_average || 1000);
  if (ratio > 2.5) {
    parts.push(`the transaction amount of ${input.currency || '₹'}${input.amount.toLocaleString()} is ${ratio.toFixed(1)}× the customer's historical average of ${input.currency || '₹'}${input.historical_average.toLocaleString()}`);
  }
  if (input.account_age_days < 7) {
    parts.push(`the account was created only ${input.account_age_days} day(s) ago, limiting the availability of historical behavior data`);
  }

  const triggeredCount = signals.filter((s) => s.triggered).length;

  let story = 'This transaction deviates from the customer\'s established behavioral baseline. ';
  if (parts.length > 0) {
    story += parts.join(', ') + '. ';
  }
  story += `In combination, ${triggeredCount} risk signal(s) were triggered, producing a composite pattern that may indicate ${threat.toLowerCase()}. `;
  story += 'This analysis is a recommendation based on synthetic behavioral signals and should be reviewed by a human analyst before any action is taken. ';

  return story;
}

function generateTimeline(input: RiskEngineInput, finalScore: number): TimelineEvent[] {
  const events: TimelineEvent[] = [];
  const baseTime = input.timestamp ? new Date(input.timestamp) : new Date();
  let t = new Date(baseTime.getTime() - 6 * 60 * 1000);

  const fmt = (d: Date) =>
    d.toLocaleTimeString('en-US', { hour: '2-digit', minute: '2-digit', hour12: true, timeZone: 'America/New_York' });

  let score = 12;

  if (input.is_new_device) {
    events.push({ time: fmt(t), label: 'New device login', description: `Device fingerprint not previously associated with this account.`, risk_score: score, type: 'warning' });
    t = new Date(t.getTime() + 60 * 1000);
    score += 16;
  }

  if (input.is_new_location) {
    events.push({ time: fmt(t), label: 'New location detected', description: `Transaction originated from ${input.location || 'an unrecognized geographic region'}.`, risk_score: score, type: 'warning' });
    t = new Date(t.getTime() + 60 * 1000);
    score += 13;
  }

  if (input.failed_attempts >= 3) {
    events.push({ time: fmt(t), label: 'Failed authentication', description: `${input.failed_attempts} failed login attempts before success.`, risk_score: score, type: 'critical' });
    t = new Date(t.getTime() + 60 * 1000);
    score += 14;
  }

  const avg = input.historical_average || 1000;
  const smallAmt = Math.round(avg * 0.5);
  events.push({ time: fmt(t), label: `Initial transaction`, description: `${input.currency || '₹'}${smallAmt.toLocaleString()} transaction processed.`, risk_score: score, type: 'info' });
  t = new Date(t.getTime() + 60 * 1000);
  score += 9;

  if (input.transaction_velocity >= 3) {
    events.push({ time: fmt(t), label: 'Velocity increasing', description: `${input.transaction_velocity} transactions within 10 minutes — above normal baseline.`, risk_score: score, type: 'warning' });
    t = new Date(t.getTime() + 60 * 1000);
    score += 15;
  }

  if (input.amount > avg * 2.5) {
    events.push({ time: fmt(t), label: 'High-value transaction', description: `${input.currency || '₹'}${input.amount.toLocaleString()} — ${((input.amount / avg)).toFixed(1)}× historical average.`, risk_score: score, type: 'critical' });
    t = new Date(t.getTime() + 60 * 1000);
    score += 12;
  }

  events.push({ time: fmt(t), label: 'Risk engine triggered', description: 'Risk Sentinel engine evaluated all signals and computed composite score.', risk_score: finalScore, type: 'critical' });
  t = new Date(t.getTime() + 30 * 1000);

  const action = recommendAction(classifyRiskLevel(finalScore));
  events.push({ time: fmt(t), label: `AI recommends ${action}`, description: `Based on the composite risk score of ${finalScore}/100, the AI investigation agent recommends ${action}.`, risk_score: finalScore, type: action === 'APPROVE' ? 'success' : 'critical' });

  return events;
}

function generateFingerprint(input: RiskEngineInput, signals: RiskSignal[]): BehavioralFingerprint[] {
  const avg = input.historical_average || 1000;
  const amountDeviation = input.amount > avg ? Math.min(((input.amount - avg) / avg) * 100, 100) : 0;
  const velocityDeviation = Math.min(input.transaction_velocity * 15, 100);
  const failedDeviation = Math.min(input.failed_attempts * 25, 100);

  const status = (d: number): BehavioralFingerprint['status'] => {
    if (d >= 75) return 'critical';
    if (d >= 50) return 'high';
    if (d >= 25) return 'elevated';
    return 'normal';
  };

  return [
    {
      signal: 'Transaction Amount',
      normal_behavior: `${input.currency || '₹'}${avg.toLocaleString()}`,
      current_behavior: `${input.currency || '₹'}${input.amount.toLocaleString()}`,
      deviation: roundTo(amountDeviation, 1),
      status: status(amountDeviation),
    },
    {
      signal: 'Transaction Frequency',
      normal_behavior: '0–2 per 10 min',
      current_behavior: `${input.transaction_velocity} per 10 min`,
      deviation: roundTo(velocityDeviation, 1),
      status: status(velocityDeviation),
    },
    {
      signal: 'Device',
      normal_behavior: 'Known device',
      current_behavior: input.is_new_device ? 'New device' : 'Known device',
      deviation: input.is_new_device ? 100 : 0,
      status: input.is_new_device ? 'critical' : 'normal',
    },
    {
      signal: 'Location',
      normal_behavior: 'Known location',
      current_behavior: input.is_new_location ? (input.location || 'New location') : 'Known location',
      deviation: input.is_new_location ? 100 : 0,
      status: input.is_new_location ? 'critical' : 'normal',
    },
    {
      signal: 'Account Age',
      normal_behavior: '> 90 days',
      current_behavior: `${input.account_age_days} days`,
      deviation: roundTo(Math.min((90 - input.account_age_days) / 90 * 100, 100), 1),
      status: input.account_age_days < 7 ? 'critical' : input.account_age_days < 30 ? 'elevated' : 'normal',
    },
    {
      signal: 'Failed Attempts',
      normal_behavior: '0 attempts',
      current_behavior: `${input.failed_attempts} attempts`,
      deviation: roundTo(failedDeviation, 1),
      status: status(failedDeviation),
    },
  ];
}

export function buildAnalysis(input: RiskEngineInput): RiskAnalysis {
  const result = analyzeRisk(input);
  return {
    transaction_id: input.transaction_id || 'unknown',
    ...result,
  };
}

File: backend/services/risk_engine.py
"""
Transparent, explainable risk scoring engine for Risk Sentinel.

7 weighted signals producing a 0-100 risk score.
This is a deterministic, explainable engine — not a black box.
"""
import math
from typing import Any

SIGNAL_WEIGHTS = {
    "transaction_amount_anomaly": 22,
    "transaction_velocity": 18,
    "new_device": 14,
    "new_location": 12,
    "account_age": 10,
    "failed_attempts": 14,
    "historical_behavior_deviation": 10,
}
MAX_SCORE = sum(SIGNAL_WEIGHTS.values())  # 100


def clamp(n: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, n))


def round_to(n: float, decimals: int = 0) -> float:
    factor = 10 ** decimals
    return round(n * factor) / factor


def classify_risk_level(score: float) -> str:
    if score <= 30:
        return "LOW"
    if score <= 60:
        return "MEDIUM"
    if score <= 80:
        return "HIGH"
    return "CRITICAL"


def recommend_action(level: str) -> str:
    return {
        "LOW": "APPROVE",
        "MEDIUM": "REVIEW",
        "HIGH": "HOLD",
        "CRITICAL": "HOLD",
    }[level]


def analyze_risk(data: dict[str, Any]) -> dict[str, Any]:
    """Analyze transaction risk and return full risk analysis."""
    amount = float(data["amount"])
    currency = data.get("currency", "₹")
    account_age_days = int(data["account_age_days"])
    is_new_device = bool(data["is_new_device"])
    is_new_location = bool(data["is_new_location"])
    failed_attempts = int(data["failed_attempts"])
    transaction_velocity = int(data["transaction_velocity"])
    historical_average = float(data.get("historical_average", 1000)) or 1000

    signals: list[dict] = []

    # 1. Transaction amount anomaly
    amount_ratio = amount / historical_average
    amount_anomaly_score = clamp((amount_ratio - 1) * 30, 0, 100) if amount_ratio > 1 else 0
    amount_triggered = amount_ratio > 2.5
    signals.append({
        "key": "transaction_amount_anomaly",
        "label": "Transaction Amount Anomaly",
        "weight": SIGNAL_WEIGHTS["transaction_amount_anomaly"],
        "triggered": amount_triggered,
        "value": f"{currency}{int(amount):,}",
        "normal_value": f"{currency}{int(historical_average):,}",
        "deviation": round_to(amount_anomaly_score, 1),
    })

    # 2. Transaction velocity
    velocity_score = clamp(transaction_velocity * 12, 0, 100)
    velocity_triggered = transaction_velocity >= 5
    signals.append({
        "key": "transaction_velocity",
        "label": "Transaction Velocity",
        "weight": SIGNAL_WEIGHTS["transaction_velocity"],
        "triggered": velocity_triggered,
        "value": f"{transaction_velocity} txns / 10 min",
        "normal_value": "0–2 txns / 10 min",
        "deviation": round_to(velocity_score, 1),
    })

    # 3. New device
    signals.append({
        "key": "new_device",
        "label": "New Device",
        "weight": SIGNAL_WEIGHTS["new_device"],
        "triggered": is_new_device,
        "value": "New device detected" if is_new_device else "Known device",
        "normal_value": "Known device",
        "deviation": 100.0 if is_new_device else 0.0,
    })

    # 4. New location
    location_str = data.get("location", "New location")
    signals.append({
        "key": "new_location",
        "label": "New Location",
        "weight": SIGNAL_WEIGHTS["new_location"],
        "triggered": is_new_location,
        "value": location_str if is_new_location else "Known location",
        "normal_value": "Known location",
        "deviation": 100.0 if is_new_location else 0.0,
    })

    # 5. Account age
    age_score = clamp((30 - account_age_days) * 3.3, 0, 100)
    age_triggered = account_age_days < 7
    signals.append({
        "key": "account_age",
        "label": "Account Age",
        "weight": SIGNAL_WEIGHTS["account_age"],
        "triggered": age_triggered,
        "value": f"{account_age_days} days",
        "normal_value": "> 90 days",
        "deviation": round_to(age_score, 1),
    })

    # 6. Failed attempts
    failed_score = clamp(failed_attempts * 20, 0, 100)
    failed_triggered = failed_attempts >= 3
    signals.append({
        "key": "failed_attempts",
        "label": "Failed Authentication Attempts",
        "weight": SIGNAL_WEIGHTS["failed_attempts"],
        "triggered": failed_triggered,
        "value": f"{failed_attempts} attempts",
        "normal_value": "0 attempts",
        "deviation": round_to(failed_score, 1),
    })

    # 7. Historical behavior deviation (composite)
    composite = (
        amount_anomaly_score * 0.3
        + velocity_score * 0.25
        + (100 if is_new_device else 0) * 0.15
        + (100 if is_new_location else 0) * 0.15
        + age_score * 0.05
        + failed_score * 0.1
    )
    hist_triggered = composite > 50
    signals.append({
        "key": "historical_behavior_deviation",
        "label": "Historical Behavior Deviation",
        "weight": SIGNAL_WEIGHTS["historical_behavior_deviation"],
        "triggered": hist_triggered,
        "value": f"{round_to(composite, 1)}% deviation",
        "normal_value": "< 20% deviation",
        "deviation": round_to(composite, 1),
    })

    # Weighted score
    raw_score = sum((s["deviation"] / 100) * s["weight"] for s in signals)
    risk_score = int(clamp(round_to((raw_score / MAX_SCORE) * 100, 0), 0, 100))
    risk_level = classify_risk_level(risk_score)
    recommended_action = recommend_action(risk_level)

    threat_type = _classify_threat(data, signals)
    attack_story = _generate_attack_story(data, signals, threat_type)
    timeline = _generate_timeline(data, risk_score)
    behavioral_fingerprint = _generate_fingerprint(data)
    signal_breakdown = {s["key"]: round(s["deviation"] * s["weight"] / 100, 1) for s in signals}

    return {
        "risk_score": risk_score,
        "risk_level": risk_level,
        "threat_type": threat_type,
        "recommended_action": recommended_action,
        "risk_factors": signals,
        "attack_story": attack_story,
        "timeline": timeline,
        "behavioral_fingerprint": behavioral_fingerprint,
        "signal_breakdown": signal_breakdown,
    }


def _classify_threat(data: dict, signals: list[dict]) -> str:
    is_new_device = bool(data["is_new_device"])
    is_new_location = bool(data["is_new_location"])
    failed = int(data["failed_attempts"]) >= 3
    high_velocity = int(data["transaction_velocity"]) >= 5
    young_account = int(data["account_age_days"]) < 7
    high_amount = float(data["amount"]) / (float(data.get("historical_average", 1000)) or 1000) > 3

    triggered = [s["key"] for s in signals if s["triggered"]]

    if not triggered or (len(triggered) == 1 and triggered[0] == "account_age" and not young_account):
        return "Normal Transaction"

    if is_new_device and is_new_location and failed:
        return "Possible Account Takeover"
    if high_velocity and high_amount:
        return "Velocity Attack"
    if is_new_device and not is_new_location and not failed:
        return "Suspicious Device"
    if high_amount and not is_new_device and not is_new_location:
        return "Unusual Amount"
    if failed and not is_new_device and not is_new_location:
        return "Multiple Failed Attempts"
    if young_account and (is_new_device or is_new_location):
        return "Potential Synthetic Identity"

    if is_new_device and is_new_location:
        return "Possible Account Takeover"
    if high_velocity:
        return "Velocity Attack"
    if is_new_device:
        return "Suspicious Device"

    return "Unusual Amount"


def _generate_attack_story(data: dict, signals: list[dict], threat: str) -> str:
    if threat == "Normal Transaction":
        return ("This transaction appears consistent with the customer's historical behavior. "
                "No significant anomalies were detected across device, location, amount, or velocity signals. "
                "The risk engine recommends approval with standard monitoring.")

    parts: list[str] = []
    currency = data.get("currency", "₹")
    amount = float(data["amount"])
    historical_average = float(data.get("historical_average", 1000)) or 1000

    if data["is_new_device"]:
        parts.append("A previously unseen device was used to initiate this transaction")
    if data["is_new_location"]:
        parts.append(f"the transaction originated from a new geographic location ({data.get('location', 'unrecognized region')})")
    if int(data["failed_attempts"]) >= 3:
        parts.append(f"{data['failed_attempts']} failed authentication attempts preceded the successful transaction")
    if int(data["transaction_velocity"]) >= 5:
        parts.append(f"{data['transaction_velocity']} transactions were attempted within a short 10-minute window, significantly above the expected baseline")
    ratio = amount / historical_average
    if ratio > 2.5:
        parts.append(f"the transaction amount of {currency}{int(amount):,} is {ratio:.1f}× the customer's historical average of {currency}{int(historical_average):,}")
    if int(data["account_age_days"]) < 7:
        parts.append(f"the account was created only {data['account_age_days']} day(s) ago, limiting the availability of historical behavior data")

    triggered_count = sum(1 for s in signals if s["triggered"])

    story = "This transaction deviates from the customer's established behavioral baseline. "
    if parts:
        story += ", ".join(parts) + ". "
    story += f"In combination, {triggered_count} risk signal(s) were triggered, producing a composite pattern that may indicate {threat.lower()}. "
    story += "This analysis is a recommendation based on synthetic behavioral signals and should be reviewed by a human analyst before any action is taken. "

    return story


def _format_time(dt) -> str:
    """Format datetime to 12-hour time string."""
    from datetime import timezone
    try:
        import zoneinfo
        tz = zoneinfo.ZoneInfo("America/New_York")
    except Exception:
        tz = timezone.utc
    return dt.astimezone(tz).strftime("%I:%M %p")


def _generate_timeline(data: dict, final_score: int) -> list[dict]:
    from datetime import datetime, timedelta, timezone

    ts = data.get("timestamp")
    if ts:
        try:
            base_time = datetime.fromisoformat(ts.replace("Z", "+00:00"))
        except Exception:
            base_time = datetime.now(timezone.utc)
    else:
        base_time = datetime.now(timezone.utc)

    t = base_time - timedelta(minutes=6)
    events: list[dict] = []
    score = 12
    currency = data.get("currency", "₹")

    if data["is_new_device"]:
        events.append({"time": _format_time(t), "label": "New device login", "description": "Device fingerprint not previously associated with this account.", "risk_score": score, "type": "warning"})
        t += timedelta(minutes=1)
        score += 16

    if data["is_new_location"]:
        events.append({"time": _format_time(t), "label": "New location detected", "description": f"Transaction originated from {data.get('location', 'an unrecognized geographic region')}.", "risk_score": score, "type": "warning"})
        t += timedelta(minutes=1)
        score += 13

    if int(data["failed_attempts"]) >= 3:
        events.append({"time": _format_time(t), "label": "Failed authentication", "description": f"{data['failed_attempts']} failed login attempts before success.", "risk_score": score, "type": "critical"})
        t += timedelta(minutes=1)
        score += 14

    avg = float(data.get("historical_average", 1000)) or 1000
    small_amt = round(avg * 0.5)
    events.append({"time": _format_time(t), "label": "Initial transaction", "description": f"{currency}{small_amt:,} transaction processed.", "risk_score": score, "type": "info"})
    t += timedelta(minutes=1)
    score += 9

    if int(data["transaction_velocity"]) >= 3:
        events.append({"time": _format_time(t), "label": "Velocity increasing", "description": f"{data['transaction_velocity']} transactions within 10 minutes — above normal baseline.", "risk_score": score, "type": "warning"})
        t += timedelta(minutes=1)
        score += 15

    if float(data["amount"]) > avg * 2.5:
        events.append({"time": _format_time(t), "label": "High-value transaction", "description": f"{currency}{int(data['amount']):,} — {(float(data['amount']) / avg):.1f}× historical average.", "risk_score": score, "type": "critical"})
        t += timedelta(minutes=1)
        score += 12

    events.append({"time": _format_time(t), "label": "Risk engine triggered", "description": "Risk Sentinel engine evaluated all signals and computed composite score.", "risk_score": final_score, "type": "critical"})
    t += timedelta(seconds=30)

    action = recommend_action(classify_risk_level(final_score))
    events.append({"time": _format_time(t), "label": f"AI recommends {action}", "description": f"Based on the composite risk score of {final_score}/100, the AI investigation agent recommends {action}.", "risk_score": final_score, "type": "success" if action == "APPROVE" else "critical"})

    return events


def _generate_fingerprint(data: dict) -> list[dict]:
    currency = data.get("currency", "₹")
    avg = float(data.get("historical_average", 1000)) or 1000
    amount = float(data["amount"])

    amount_deviation = min(((amount - avg) / avg) * 100, 100) if amount > avg else 0
    velocity_deviation = min(int(data["transaction_velocity"]) * 15, 100)
    failed_deviation = min(int(data["failed_attempts"]) * 25, 100)

    def status(d: float) -> str:
        if d >= 75:
            return "critical"
        if d >= 50:
            return "high"
        if d >= 25:
            return "elevated"
        return "normal"

    account_age_days = int(data["account_age_days"])
    age_dev = round_to(min((90 - account_age_days) / 90 * 100, 100), 1)
    age_status = "critical" if account_age_days < 7 else "elevated" if account_age_days < 30 else "normal"

    return [
        {"signal": "Transaction Amount", "normal_behavior": f"{currency}{int(avg):,}", "current_behavior": f"{currency}{int(amount):,}", "deviation": round_to(amount_deviation, 1), "status": status(amount_deviation)},
        {"signal": "Transaction Frequency", "normal_behavior": "0–2 per 10 min", "current_behavior": f"{data['transaction_velocity']} per 10 min", "deviation": round_to(velocity_deviation, 1), "status": status(velocity_deviation)},
        {"signal": "Device", "normal_behavior": "Known device", "current_behavior": "New device" if data["is_new_device"] else "Known device", "deviation": 100.0 if data["is_new_device"] else 0.0, "status": "critical" if data["is_new_device"] else "normal"},
        {"signal": "Location", "normal_behavior": "Known location", "current_behavior": data.get("location", "New location") if data["is_new_location"] else "Known location", "deviation": 100.0 if data["is_new_location"] else 0.0, "status": "critical" if data["is_new_location"] else "normal"},
        {"signal": "Account Age", "normal_behavior": "> 90 days", "current_behavior": f"{account_age_days} days", "deviation": age_dev, "status": age_status},
        {"signal": "Failed Attempts", "normal_behavior": "0 attempts", "current_behavior": f"{data['failed_attempts']} attempts", "deviation": round_to(failed_deviation, 1), "status": status(failed_deviation)},
    ]
